File: android/app/src/main/kotlin/studio/maximumimpact/tokencounter/notifications/SpendAlert.kt
package studio.maximumimpact.tokencounter.notifications

import java.time.LocalDate

/**
 * Pure decision logic for the spend alert — no Android deps, so it's unit
 * testable. The worker combines [atThreshold] with a once-per-[monthKey] dedupe.
 */
object SpendAlert {

    /** Fraction of the limit at which we notify. */
    const val THRESHOLD_FRACTION = 0.90

    /** True when spend has reached >= 90% of a positive limit. */
    fun atThreshold(spentCents: Long, limitCents: Long): Boolean {
        if (limitCents <= 0L) return false
        return spentCents.toDouble() / limitCents.toDouble() >= THRESHOLD_FRACTION
    }

    /** "yyyy-MM" used to dedupe alerts to at most once per calendar month. */
    fun monthKey(date: LocalDate): String = date.toString().substring(0, 7)
}

File: ios/TokenTracker/Core/SpendAlert.swift
import Foundation

/// Pure decision logic for the 90%-of-limit spend alert — no Android/iOS deps,
/// so it's unit testable. Swift sibling of the Android `SpendAlert` object. The
/// background check combines [atThreshold] with a once-per-[monthKey] dedupe.
enum SpendAlert {

    /// Fraction of the limit at which we notify.
    static let thresholdFraction = 0.90

    /// True when spend has reached ≥ 90% of a positive limit.
    static func atThreshold(spentCents: Int64, limitCents: Int64) -> Bool {
        guard limitCents > 0 else { return false }
        return Double(spentCents) / Double(limitCents) >= thresholdFraction
    }

    /// "yyyy-MM" (UTC) used to dedupe alerts to at most once per calendar month.
    static func monthKey(_ date: Date) -> String {
        var calendar = Calendar(identifier: .iso8601)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        let comps = calendar.dateComponents([.year, .month], from: date)
        return String(format: "%04d-%02d", comps.year ?? 0, comps.month ?? 0)
    }
}
